### PR TITLE
Convert int to roman

### PR DESCRIPTION
**Pull request by KimaciaJnr**
---
name: solution review 
about: A template PR for code review with a checklist
---

# Behavior

## Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [ ] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called in the function file

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/convert_int_to_roman [hunk 1]
```diff
@@ -0,0 +1,51 @@
+#!/usr/bin/env python3
```
Comment (AmmarIbrahimTech) [inline solutions/convert_int_to_roman]: ## Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file

complete the module docstring like this:
```
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
A module for converting integers to roman.

Module contents:
    - convert_int_to_roman: converts numbers from integer to roman.

Created on
@author: 
"""
```

File: solutions/tests/test_convert_int_to_roman [hunk 2]
```diff
@@ -0,0 +1,56 @@
+#!/usr/bin/env python3
```
Comment (AmmarIbrahimTech) [inline solutions/tests/test_convert_int_to_roman]: - [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the test file:

complete the module docstring like this:
```
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

# -*- coding: utf-8 -*-
"""
Unit test for convert_int_to_roman function.

Module contents:
    - TestConvertIntToRoman: Unit test class for the convert_int_to_roman function.

Created on 
@author:
"""
```

File: solutions/tests/test_convert_int_to_roman [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+Created on 04/01/2025
+@author: Peter Ngugi
+"""
+
+import unittest
+
+from mit_solutions import convert_int_to_roman
+
+
+class TestConvertIntToRoman(unittest.TestCase):
```
Comment (AmmarIbrahimTech) [inline solutions/tests/test_convert_int_to_roman]: ### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring

Perfecto

File: solutions/tests/test_convert_int_to_roman [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    """Test the Convert_int_to_roman function"""
+
+    def test_single_digit_int(self):
```
Comment (AmmarIbrahimTech) [inline solutions/tests/test_convert_int_to_roman]: - Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
  
  Great job. Also, the ```actual``` & ```expected``` variables made it readable and easy to debug. 👏

File: solutions/tests/test_convert_int_to_roman [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+        self.assertEqual(actual, expected)
+
+    def test_triple_digit_int(self):
+        """Should return a roman numeral
+        representation for a triple-digit integer"""
+        actual = convert_int_to_roman(387)
+        expected = "CCCLXXXVII"
+        self.assertEqual(actual, expected)
+
+    def test_boundary_value(self):
```
Comment (AmmarIbrahimTech) [inline solutions/tests/test_convert_int_to_roman]: - [x] There are tests for boundary cases

File: solutions/tests/test_convert_int_to_roman [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+        """Should return a roman numeral
+        representation for the maximum input of 3999"""
+        actual = convert_int_to_roman(3999)
+        expected = "MMMCMXCIX"
+        self.assertEqual(actual, expected)
+
+    def test_invalid_input_zero(self):
```
Comment (AmmarIbrahimTech) [inline solutions/tests/test_convert_int_to_roman]: - [x] There are tests for defensive assertions

File: solutions/tests/test_convert_int_to_roman
Comment (AmmarIbrahimTech) [inline solutions/tests/test_convert_int_to_roman:1]: - [ ] All tests pass

The tests pass the CI checks (I don't know why) but they don't pass in the local device and codespace because there are two problems that lead to ImportError 

- The first is that the file names don't contain ```.py```
suggestion: add the ```.py``` extensions to your solution and test files

   - make them like this:
      ```convert_int_to_roman.py```
      ```test_convert_int_to_roman.py```


- The 2nd problem is that line 10 ```from mit_solutions import convert_int_to_roman``` doesn't contain the right path for the module from which the function will be imported
   - the correct path is:
   ```from ..convert_int_to_roman import convert_int_to_roman```

File: solutions/convert_int_to_roman [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    Examples:
+    >>> int_to_roman(5)
+        "V"
+    >>> int_to_roman(100)
+        "C"
+    >>> int_to_roman(33)
+        "XXXIII"
+    >>> int_to_roman(1)
+        'I'
+    """
```
Comment (AmmarIbrahimTech) [inline solutions/convert_int_to_roman]: ### Function Docstring
There are 2 minor problems in the docstring syntax but they are leading to doctest failure.

I will try to explain them one by one and provide the suggested correction for them (to be clear, in each problem I will provide suggestion for just fixing it. And I will keep the other problems unfixed to fix them in their point. This way It will be clear how to fix each problem and what eror is this problem leading to)

1) Function name call in the examples 
```
Examples:
 >>> int_to_roman(5)
 ```
 
 - What is wrong?
The function name is written as ```int_to_roman``` , where the real name is ```convert_int_to_roman```

- Eror this issue is leading to:
```NameError: name 'int_to_roman' is not defined```

- How to fix?
rename the function in the examples with the real function name
```
Examples:
    >>> convert_int_to_roman(5)
        "V"
    >>> convert_int_to_roman(100)
        "C"
    >>> convert_int_to_roman(33)
        "XXXIII"
    >>> convert_int_to_roman(1)
        'I'
 ```

2) indentation and string formating:
 
- What is wrong?
   1) there are indentations in the expected values lines that shouldn't be there in the doctest format
   2) the string mark ```" "``` is written in the double mark way ```"V"``` , while doctest is expecting it to be single mark ```'V'```
   {it leads to eror in my vs code but I think logically it shouldn't so just keep it in mind and try to run your doctest ``python -m doctest -v .\solutions\convert_int_to_roman.py``` and see if you will get the same following error)

- Eror this issue is leading to:
```
Failed example:
    convert_int_to_roman(5)
Expected:
    "V"
Got:
    'V'
Trying:
    convert_int_to_roman(100)
Expecting:
    "C"
```
- How to fix?
delete the indentation before the expected values and replace the ``` " "``` with ``` ' ' ```

```
Examples:
    >>> convert_int_to_roman(5)
    'V'
    >>> convert_int_to_roman(100)
    'C'
    >>> convert_int_to_roman(33)
    'XXXIII'
    >>> convert_int_to_roman(1)
    'I'
    """
```

File: solutions/convert_int_to_roman
Comment (AmmarIbrahimTech) [inline solutions/convert_int_to_roman:1]: ### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called in the function file

File: solutions/convert_int_to_roman [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+        50: "L",
+        90: "XC",
+        100: "C",
+        400: "CD",
+        500: "D",
+        900: "CM",
+        1000: "M",
+    }
+    r = ""
+    values = sorted(romans.keys(), reverse=True)
```
Comment (AmmarIbrahimTech) [inline solutions/convert_int_to_roman]: - [x] Variable names are clear and helpful

File: solutions/convert_int_to_roman [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+    >>> int_to_roman(5)
+        "V"
+    >>> int_to_roman(100)
+        "C"
+    >>> int_to_roman(33)
+        "XXXIII"
+    >>> int_to_roman(1)
+        'I'
+    """
+    if not (1 <= num <= 3999):
```
Comment (AmmarIbrahimTech) [inline solutions/convert_int_to_roman]: - [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

**Review by AmmarIbrahimTech — CHANGES_REQUESTED** [review]
The overall code is more than great.

I have added some suggestions for improving it and to fix some tests erors

File: solutions/convert_int_to_roman (hunk 1 above)
Comment (KimaciaJnr) [reply solutions/convert_int_to_roman]: Done!

File: solutions/tests/test_convert_int_to_roman (hunk 2 above)
Comment (KimaciaJnr) [reply solutions/tests/test_convert_int_to_roman]: Done!

File: solutions/tests/test_convert_int_to_roman
Comment (KimaciaJnr) [reply solutions/tests/test_convert_int_to_roman:1]: Done!

File: solutions/convert_int_to_roman (hunk 7 above)
Comment (KimaciaJnr) [reply solutions/convert_int_to_roman]: Done!

**Comment by AmmarIbrahimTech** [comment]
https://github.com/user-attachments/assets/fab31bd6-1fcc-40e8-9ab5-37ef96bfb0be

**Comment by KimaciaJnr** [comment]
> Doctest.failure.explinasion.mp4

Ave changed the double quotes to single quotes.

**Comment by njeridorcaswanja** [comment]
Great work on this. The code is clean and easy to follow. I particularly like how you've handled edge cases and provided detailed test cases.

**Comment by AmmarIbrahimTech** [comment]
> > Doctest.failure.explinasion.mp4
> 
> Ave changed the double quotes to single quotes.

Great work. Now the test pass 👏 

image

**Review by AmmarIbrahimTech — APPROVED** [review]
Done ✔️ (Ready to merge)
